README.md
# leia-fun


ChangeLog.md
# Changelog for leia-fun

## Unreleased changes


src/Lib.hs
module Lib
    ( someFunc
    , elf
    , reindeer
    , newGroup
    , santa
    , awhile
    , testDelay
    , setupLog
    ) where

import Control.Concurrent.STM
import Control.Concurrent
import System.Random
import System.Log.Caster

someFunc :: IO ()
someFunc = putStrLn "someFunc"

data Gate = MkGate Int (TVar Int)

newGate :: Int -> STM Gate
newGate n = do
    tv <- newTVar 0
    return (MkGate n tv)

passGate :: Gate -> IO ()
passGate (MkGate n tv)
  = atomically (do n_left <- readTVar tv
                   check (n_left > 0)
                   writeTVar tv (n_left-1))

operateGate :: Gate -> IO ()
operateGate (MkGate n tv) = do
    atomically (writeTVar tv n)
    atomically (do n_left <- readTVar tv
                   check (n_left == 0))

data Group = MkGroup Int (TVar (Int, Gate, Gate))

newGroup   :: Int -> IO Group
newGroup n = atomically (do g1 <- newGate n; g2 <- newGate n
                            tv <- newTVar (n, g1, g2)
                            return (MkGroup n tv))

joinGroup  :: Group -> IO (Gate,Gate)
joinGroup (MkGroup n tv)
  = atomically (do (n_left, g1, g2) <- readTVar tv
                   check (n_left > 0)
                   writeTVar tv (n_left-1, g1, g2)
                   return (g1,g2))

awaitGroup :: Group -> STM (Gate,Gate)
awaitGroup (MkGroup n tv) = do
    (n_left, g1, g2) <- readTVar tv
    check (n_left == 0)
    new_g1 <- newGate n; new_g2 <- newGate n
    writeTVar tv (n,new_g1,new_g2)
    return (g1,g2)

meetInStudy :: LogQueue -> Int -> IO ()
meetInStudy lq id = info lq $ "Elf " ++ show id ++ " meeting in the study"

deliverToys :: LogQueue -> Int -> IO ()
deliverToys lq id = info lq $ "Reindeer " ++ show id ++ " delivering toys"

helper1 :: LogQueue -> Group -> IO () -> IO ()
helper1 lq group do_task = do
    (in_gate, out_gate) <- joinGroup group
    passGate in_gate
    randomDelay lq 1
    info lq "Doing task"
    do_task
    info lq "Task done"
    passGate out_gate

elf1, reindeer1 :: LogQueue -> Group -> Int -> IO ()
elf1      lq gp id = helper1 lq gp (meetInStudy lq id)
reindeer1 lq gp id = helper1 lq gp (deliverToys lq id)

elf :: LogQueue -> Group -> Int -> IO ThreadId
elf lq gp id = forkIO (awhile (do randomDelay lq 10
                                  elf1 lq gp id))

reindeer :: LogQueue -> Group -> Int -> IO ThreadId
reindeer lq gp id = forkIO (awhile (do randomDelay lq 10
                                       reindeer1 lq gp id))

santa :: LogQueue -> Group -> Group -> IO ()
santa lq elf_gp rein_gp = do
    info lq  ">>>"
    (task, (in_gate, out_gate)) <- atomically (orElse
                     (chooseGroup rein_gp "deliver toys")
                     (chooseGroup elf_gp "meet in my study"))
    info lq $ "Ho! Ho! Ho! let’s " ++ task
    operateGate in_gate
              -- Now the helpers do their task
    operateGate out_gate
    info lq  "<<<"
  where
    chooseGroup :: Group -> String -> STM (String, (Gate,Gate))
    chooseGroup gp task = do gates <- awaitGroup gp
                             return (task, gates)

awhile :: IO () -> IO ()
-- Repeatedly perform the action
awhile act = forever' act 10
  where -- cheating here to make it stop eventually
    forever' :: IO () -> Int -> IO ()
    forever' act 0 = return ()
    forever' act n = do
        act
        forever' act (n - 1)

randomDelay :: LogQueue -> Int -> IO ()
-- Delay for a random time between 1 and 1,000,000 microseconds
randomDelay lq seconds = do
  waitTime <- getStdRandom (randomR (1, seconds * 1000000))
--  info lq $ "Pause: " ++ show waitTime
  threadDelay waitTime
--  info lq $ "Resume: " ++ show waitTime

randomOp :: IO String
randomOp = do
    n <- getStdRandom (randomR (0, 3))
    return (["+", "-", "*", "/"] !! n)

testDelay :: LogQueue -> IO ()
testDelay lq = do
    info lq "All our operators are currently busy..."
    randomDelay lq 5
    op <- randomOp
    info lq ("Operator (" ++ op ++ ") here, how may I help you?")

setupLog :: IO (LogQueue, IO ())
setupLog = do
    chan <- newLogChan
    lq <- newLogQueue

    relayThreadId <- forkIO $ do
          relayLog chan LogInfo stdoutListener
          putStrLn "End relayLog"

    broadcastThreadId <- forkIO $ do
          broadcastLog lq chan
          putStrLn "End broadcastLog"

    info lq "----------"
    return (lq, threadDelay 100000)


src/TryFused.hs
{-# LANGUAGE DeriveAnyClass #-}
{-# LANGUAGE DeriveFunctor #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DerivingStrategies #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE UndecidableInstances #-}

-- http://hackage.haskell.org/package/fused-effects
-- https://www.youtube.com/watch?v=vfDazZfxlNs "Building Haskell Programs with Fused Effects" by Patrick Thomson
-- https://github.com/fused-effects/fused-effects/blob/master/examples/Teletype.hs

module TryFused
( runTeletypeIO
, tryFused
, write
) where

import Prelude hiding (read)

import Control.Effect
import Control.Effect.Carrier
import Control.Effect.State
import Control.Effect.Writer
import Control.Monad.IO.Class
import GHC.Generics (Generic1)
import System.Log.Caster

tryFused :: LogQueue -> IO ()
tryFused lq = runM . runTeletypeIO $ do
  tryX
  info lq "there!"

tryX :: (MonadIO m, Carrier sig m) => TeletypeIOC m ()
tryX = write "Hi"

data Teletype m k
  = Read (String -> m k)
  | Write String (m k)
  deriving stock (Functor, Generic1)
  deriving anyclass (HFunctor, Effect)

read :: (Member Teletype sig, Carrier sig m) => m String
read = send (Read pure)

write :: (Member Teletype sig, Carrier sig m) => String -> m ()
write s = send (Write s (pure ()))

runTeletypeIO :: TeletypeIOC m a -> m a
runTeletypeIO = runTeletypeIOC

newtype TeletypeIOC m a = TeletypeIOC { runTeletypeIOC :: m a }
  deriving newtype (Applicative, Functor, Monad, MonadIO)

instance (MonadIO m, Carrier sig m) => Carrier (Teletype :+: sig) (TeletypeIOC m) where
  eff (L (Read    k)) = liftIO getLine      >>= k
  eff (L (Write s k)) = liftIO (putStrLn s) >>  k
  eff (R other)       = TeletypeIOC (eff (handleCoercible other))


runTeletypeRet :: [String] -> TeletypeRetC m a -> m ([String], ([String], a))
runTeletypeRet i = runWriter . runState i . runTeletypeRetC

newtype TeletypeRetC m a = TeletypeRetC { runTeletypeRetC :: StateC [String] (WriterC [String] m) a }
  deriving newtype (Applicative, Functor, Monad)

instance (Carrier sig m, Effect sig) => Carrier (Teletype :+: sig) (TeletypeRetC m) where
  eff (L (Read    k)) = do
    i <- TeletypeRetC get
    case i of
      []  -> k ""
      h:t -> TeletypeRetC (put t) *> k h
  eff (L (Write s k)) = TeletypeRetC (tell [s]) *> k
  eff (R other)       = TeletypeRetC (eff (R (R (handleCoercible other))))

src/LEIA/Logging.hs
module LEIA.Logging
    ( withLog
    ) where

import Control.Concurrent
import Control.Effect
import Control.Effect.Carrier
import Control.Effect.State
import Control.Effect.Writer
import Control.Monad
import Control.Monad.IO.Class
import GHC.Generics (Generic1)
import qualified System.Log.Caster as DistCaster

setupLog :: IO (DistCaster.LogQueue, IO ())
setupLog = do
    (lq, chan) <- liftIO startLogging

    DistCaster.info lq "----------"
    return (lq, threadDelay 100000)

withLog :: (DistCaster.LogQueue -> IO a) -> IO a
withLog action = do
    (lq, finalize) <- setupLog
    result <- action lq
    finalize
    return result

data LogEffect m k
  = Debug String (m k)
  | Info String (m k)
  deriving stock (Functor, Generic1)
  deriving anyclass (HFunctor, Effect)

debug :: (Member LogEffect sig, Carrier sig m) => String -> m ()
debug s = send (Debug s (pure ()))

info :: (Member LogEffect sig, Carrier sig m) => String -> m ()
info s = send (Info s (pure ()))

startLogging :: IO (DistCaster.LogQueue, DistCaster.LogChan)
startLogging = do
    chan <- liftIO DistCaster.newLogChan
    lq <- liftIO DistCaster.newLogQueue

    relayThreadId <- liftIO . forkIO $ do
          DistCaster.relayLog chan DistCaster.LogInfo DistCaster.stdoutListener
          putStrLn "End relayLog"

    broadcastThreadId <- liftIO . forkIO $ do
          DistCaster.broadcastLog lq chan
          putStrLn "End broadcastLog"
    return (lq, chan)

runLogEffect :: (MonadIO m) => LogCasterIOC m a -> m a
runLogEffect action = do
    (lq, chan) <- liftIO startLogging

    ((liftM snd) . runState (lq, chan) . runLogCasterIOC) $ do
        result <- action
        DistCaster.info lq "----------"
        liftIO $ threadDelay 100000
        return result

newtype LogCasterIOC m a = LogCasterIOC { runLogCasterIOC :: StateC (DistCaster.LogQueue, DistCaster.LogChan) m a }
  deriving newtype (Applicative, Functor, Monad, MonadIO)


src/Santa.hs
module Santa
    ( elf
    , reindeer
    , newGroup
    , santa
    , mainSanta
    , awhile
    , testDelay
    ) where

-- https://www.schoolofhaskell.com/school/advanced-haskell/beautiful-concurrency/4-the-santa-claus-problem
-- https://www.microsoft.com/en-us/research/wp-content/uploads/2016/07/mark.pdf
-- https://simonmar.github.io/bib/papers/async.pdf

import Control.Concurrent.STM
import Control.Concurrent
import System.Random
import System.Log.Caster

data Gate = MkGate Int (TVar Int)

newGate :: Int -> STM Gate
newGate n = do
    tv <- newTVar 0
    return (MkGate n tv)

passGate :: Gate -> IO ()
passGate (MkGate n tv)
  = atomically (do n_left <- readTVar tv
                   check (n_left > 0)
                   writeTVar tv (n_left-1))

operateGate :: Gate -> IO ()
operateGate (MkGate n tv) = do
    atomically (writeTVar tv n)
    atomically (do n_left <- readTVar tv
                   check (n_left == 0))

data Group = MkGroup Int (TVar (Int, Gate, Gate))

newGroup   :: Int -> IO Group
newGroup n = atomically (do g1 <- newGate n; g2 <- newGate n
                            tv <- newTVar (n, g1, g2)
                            return (MkGroup n tv))

joinGroup  :: Group -> IO (Gate,Gate)
joinGroup (MkGroup n tv)
  = atomically (do (n_left, g1, g2) <- readTVar tv
                   check (n_left > 0)
                   writeTVar tv (n_left-1, g1, g2)
                   return (g1,g2))

awaitGroup :: Group -> STM (Gate,Gate)
awaitGroup (MkGroup n tv) = do
    (n_left, g1, g2) <- readTVar tv
    check (n_left == 0)
    new_g1 <- newGate n; new_g2 <- newGate n
    writeTVar tv (n,new_g1,new_g2)
    return (g1,g2)

meetInStudy :: LogQueue -> Int -> IO ()
meetInStudy lq id = info lq $ "Elf " ++ show id ++ " meeting in the study"

deliverToys :: LogQueue -> Int -> IO ()
deliverToys lq id = info lq $ "Reindeer " ++ show id ++ " delivering toys"

helper1 :: LogQueue -> Group -> IO () -> IO ()
helper1 lq group do_task = do
    (in_gate, out_gate) <- joinGroup group
    passGate in_gate
    randomDelay lq 1
    info lq "Doing task"
    do_task
    info lq "Task done"
    passGate out_gate

elf1, reindeer1 :: LogQueue -> Group -> Int -> IO ()
elf1      lq gp id = helper1 lq gp (meetInStudy lq id)
reindeer1 lq gp id = helper1 lq gp (deliverToys lq id)

elf :: LogQueue -> Group -> Int -> IO ThreadId
elf lq gp id = forkIO (awhile (do randomDelay lq 10
                                  elf1 lq gp id))

reindeer :: LogQueue -> Group -> Int -> IO ThreadId
reindeer lq gp id = forkIO (awhile (do randomDelay lq 10
                                       reindeer1 lq gp id))

santa :: LogQueue -> Group -> Group -> IO ()
santa lq elf_gp rein_gp = do
    info lq  ">>>"
    (task, (in_gate, out_gate)) <- atomically (orElse
                     (chooseGroup rein_gp "deliver toys")
                     (chooseGroup elf_gp "meet in my study"))
    info lq $ "Ho! Ho! Ho! let’s " ++ task
    operateGate in_gate
              -- Now the helpers do their task
    operateGate out_gate
    info lq  "<<<"
  where
    chooseGroup :: Group -> String -> STM (String, (Gate,Gate))
    chooseGroup gp task = do gates <- awaitGroup gp
                             return (task, gates)

mainSanta :: LogQueue -> IO ()
mainSanta lq = do
    elf_group <- newGroup 3
    sequence_ [ elf lq elf_group n | n <- [1..10] ]

    rein_group <- newGroup 9
    sequence_ [ reindeer lq rein_group n | n <- [1..9] ]

    awhile (santa lq elf_group rein_group)

awhile :: IO () -> IO ()
-- Repeatedly perform the action
awhile act = forever' act 10
  where -- cheating here to make it stop eventually
    forever' :: IO () -> Int -> IO ()
    forever' act 0 = return ()
    forever' act n = do
        act
        forever' act (n - 1)

randomDelay :: LogQueue -> Int -> IO ()
-- Delay for a random time between 1 and 1,000,000 microseconds
randomDelay lq seconds = do
  waitTime <- getStdRandom (randomR (1, seconds * 1000000))
--  info lq $ "Pause: " ++ show waitTime
  threadDelay waitTime
--  info lq $ "Resume: " ++ show waitTime

randomOp :: IO String
randomOp = do
    n <- getStdRandom (randomR (0, 3))
    return (["+", "-", "*", "/"] !! n)

testDelay :: LogQueue -> IO ()
testDelay lq = do
    info lq "All our operators are currently busy..."
    randomDelay lq 5
    op <- randomOp
    info lq ("Operator (" ++ op ++ ") here, how may I help you?")
